[GBM] CRendererDRMPRIMEGLES improvements

lrusak opened the pull request:
This PR adds a new class `CEGLFence` which has been designed to help using EGL sync objects. This works by inserting a sync object into the command stream and then signaling that sync object when it has been rendered. This allows us to check the signaled status so we can decide if we can reuse the buffer or not. For more information please see [1].

This also updates the VBO's to the more standardized format used in other renderers.

The last commit adds proper override methods for the renderer to properly expose the features and scalers that are able to be use with this renderer.

references:
 1) https://www.khronos.org/registry/EGL/extensions/KHR/EGL_KHR_fence_sync.txt

lrusak committed:
CRendererDRMPRIMEGLES: update VBO's to be similar to CLinuxRendererGL

lrusak committed:
CRendererDRMPRIMEGLES: add override methods for ERENDERFEATURE and ESCALINGMETHOD

lrusak committed:
CEGLFence: add class to help with EGL sync objects

lrusak committed:
CRendererDRMPRIMEGLES: use CEGLFence to sync rendering

pkerling commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.cpp:
```diff
@@ -43,15 +44,32 @@ bool CRendererDRMPRIMEGLES::Configure(const VideoPicture &picture, float fps, un
 
+bool CRendererDRMPRIMEGLES::NeedBuffer(int index)
+{
+  if (m_fences[index]->IsSignaled())
```
`return !m_fences[index]->IsSignaled()`?

lrusak replied:
If you think it's more clear then sure I can do that ❤️

pkerling replied:
I just like more concise code personally, it is not "wrong" as-is by any means.

I don't always write the disclaimer, but I think you know that most of the time I'm just making personal suggestions for alternatives in which you could argue for either side :) Here I think the argument would be that, written as you did it, it is easier to extend the function with other checks should the need arise.

pkerling commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.cpp:
```diff
@@ -188,3 +222,32 @@ bool CRendererDRMPRIMEGLES::RenderHook(int index)
+
+bool CRendererDRMPRIMEGLES::Supports(ERENDERFEATURE feature)
+{
+  if (feature == RENDERFEATURE_STRETCH ||
```
same as above

these kind of functions (SupportsXXX) also often use switch/case constructs

pkerling commented on xbmc/utils/EGLFence.cpp:
```diff
@@ -0,0 +1,68 @@
+  if (m_fence == EGL_NO_SYNC_KHR)
+  {
+    CEGLUtils::LogError("failed to create egl sync fence");
+    return false;
```
What use is this return value?

lrusak replied:
you're right, I don't actually deal with it anywhere. I guess I really only need to log the failure and change bool to void?

pkerling replied:
The question that should be always asked in cases like this is: Is failure expected? I.e., under what circumstances can `eglCreateSyncKHR` fail? Going further: If it fails, can the rest of the code run as normal or will it break somewhere? If failure is expected, there may or may not be any need to log a message, I'll leave that up to you. If it is not expected, why not throw an exception?

I'm a big fan of hard failing by the way. In lots of places in the Kodi code we log a warning and continue doing *something* where we should really just be throwing an exception. Hard-failing allows you to catch bugs sooner because they don't get lost in the log noise.

lrusak replied:
I read the docs and it seems like the only times a fence cannot be created are because the EGLDisplay is wrong or the EGLContext changed. These are pretty large errors so let's hard fail for now. I could never get the create fence call to fail in my limited testing.

yol reviewed (approved):
code looks good!

You could namespace `CEGLFence` but yeah we don't do that very consistently, unfortunately

lrusak commented:
> code looks good!

Only because I learn from the master 👨‍🏫 

> You could namespace `CEGLFence` but yeah we don't do that very consistently, unfortunately

What should the namespace be? Should we have an EGL namespace for all egl related classes (CEGLImage, etc)?

pkerling commented:
> What should the namespace be? Should we have an EGL namespace for all egl related classes (CEGLImage, etc)?

per our style guide, everything should be under `KODI`, and for the rest it most of the time follows the directory layout (roughly at least). So EGL stuff would fall under `KODI::UTILS` in this system. `KODI::UTILS::EGL` also sounds fine though.

lrusak commented:
Updated with the suggested changes

After more testing I also changed it so the `IsSignaled()` will return true if `m_fence == EGL_NO_SYNC_KHR`. I did this because I found that after a seek the buffers were returned and the fences were destroyed so when the renderer tried to call `NeedBuffer()` it would the fence would never be able to be signaled because the fence was destroyed already. I'm not sure if this is the best behaviour but it seemed to work with my testing and I logged when no fence existed and when the fence was signaled and I seemed to have the desired behaviour of the fences in general.

pkerling commented on xbmc/utils/EGLFence.cpp:
```diff
@@ -0,0 +1,69 @@
+  if (m_fence == EGL_NO_SYNC_KHR)
+  {
+    CEGLUtils::LogError("failed to create egl sync fence");
+    std::runtime_error("failed to create egl sync fence");
```
```suggestion
    throw std::runtime_error("failed to create egl sync fence");
```

Genuinely surprised that you didn't get a warning here :-)

lrusak replied:
Nope, not even a warning :man_shrugging:

pkerling commented:
Merge ahead :tada:
